### PR TITLE
split_controller: fix the incorrect sampled key ranges number check

### LINKED ISSUE #12012: Incorrect sampled key ranges number check in the Load Base Split
### What version of TiKV are you using?

After merging https://github.com/tikv/tikv/pull/11986. 

### What did you expect?

The number of sampled key ranges that will be evaluated is actually the number of `self.key_ranges` flatten.

https://github.com/tikv/tikv/blob/cbcaab6ce56cdc0a503a5f9171470bff3c0b53da/components/raftstore/src/store/worker/split_controller.rs#L258-L260

### What did happened?

However, the check uses `self.key_ranges.len()` to do the check incorrectly.

https://github.com/tikv/tikv/blob/cbcaab6ce56cdc0a503a5f9171470bff3c0b53da/components/raftstore/src/store/worker/split_controller.rs#L252

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What is changed and how it works?

Issue Number: Close #12012.

What's Changed:

```commit-message
split_controller: fix the incorrect sampled key ranges number check
```

### Related changes

- Problem cause: https://github.com/tikv/tikv/pull/11986

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None.
```

### CODE REVIEW

**Review by lhy1024 — APPROVED** [review]
Do we need to add a test to avoid the similar problem?

**Review by disksing — APPROVED** [review]

File: components/raftstore/src/store/worker/split_controller.rs [hunk 1]
```diff
@@ -247,15 +247,16 @@ impl Recorder {
             |x| x,
         );
         let mut samples = Samples::from(sampled_key_ranges);
+        let recorded_key_ranges: Vec<&KeyRange> = self.key_ranges.iter().flatten().collect();
         // Because we need to observe the number of `no_enough_key` of all the actual keys,
         // so we do this check after the samples are calculated.
-        if (self.key_ranges.len() as u64) < config.sample_threshold {
+        if (recorded_key_ranges.len() as u64) < config.sample_threshold {
```
Comment (rleungx) [inline components/raftstore/src/store/worker/split_controller.rs:253]: How about using `size_hint`?

Comment (JmPotato) [reply components/raftstore/src/store/worker/split_controller.rs:253]: `size_hint`'s output is a tuple and has an `Option` inside it, I think using `collect` is more simple here.

Comment (rleungx) [reply components/raftstore/src/store/worker/split_controller.rs:253]: I think maybe you can use `.0` so that we can avoid `collect` and `into_iter`.

Comment (JmPotato) [reply components/raftstore/src/store/worker/split_controller.rs:253]: Also, after `flatten()`, the result of `size_hint()` will be `(0, None)`.

origin_img_v2_85e192ec-377a-449b-bb6e-51ce6b1730ag

**Pull request merged**